numpy/core/src/private/npy_longdouble.c

```diff
@@ -0,0 +1,97 @@
+#include <Python.h>
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_math.h"
+
+/* This is a backport of Py_SETREF */
+#define NPY_SETREF(op, op2)                      \
+    do {                                        \
+        PyObject *_py_tmp = (PyObject *)(op);   \
+        (op) = (op2);                           \
+        Py_DECREF(_py_tmp);                     \
+    } while (0)
+
+
+/* Heavily derived from PyLong_FromDouble
+ * Notably, we can't set the digits directly, so have to shift and or instead.
+ */
+PyObject *
+PyLong_FromLongDouble(npy_longdouble ldval)
+{
+    PyObject *v;
+    PyObject *l_chunk_size;
+    const int chunk_size = 30; // number of bits to extract at a time
+    npy_longdouble frac;
+    int i, ndig, expo, neg;
+    neg = 0;
+
+    if (npy_isinf(ldval)) {
+        PyErr_SetString(PyExc_OverflowError,
+                        "cannot convert longdouble infinity to integer");
+        return NULL;
+    }
+    if (npy_isnan(ldval)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "cannot convert longdouble NaN to integer");
+        return NULL;
+    }
+    if (ldval < 0.0) {
+        neg = 1;
+        ldval = -ldval;
+    }
+    frac = npy_frexpl(ldval, &expo); /* ldval = frac*2**expo; 0.0 <= frac < 1.0 */
+    v = PyLong_FromLong(0L);
+    if (v == NULL)
+        return NULL;
+    if (expo <= 0)
+        return v;
+
+    ndig = (expo-1) / chunk_size + 1;
+
+    l_chunk_size = PyLong_FromLong(chunk_size);
+    if (l_chunk_size == NULL) {
+        Py_DECREF(v);
+        return NULL;
+    }
+
+    /* Get the MSBs of the integral part of the float */
+    frac = npy_ldexpl(frac, (expo-1) % chunk_size + 1);
+    for (i = ndig; --i >= 0; ) {
+        npy_int32 chunk = (npy_int32)frac;
```

Comment:
Why not `npy_uint64` with `chunk_size` 64?

`digits` in the original code was unsigned

Reply:
Hadn't noticed that it was unsigned. Yes, could probably change to `sizeof(npy_long) * CHAR_BIT` and `npy_long`

Reply:
Changed to use `PyLong_FromUnsignedLongLong`, `npy_ulonglong`, and `NPY_BITSOF_ULONGLONG`